### PR TITLE
Rebase for merge into upstream

### PR DESCRIPTION
**Pull request by Duologic**

### CODE REVIEW

**Comment by xordoquy** [comment]
didn't had the time to review this. Should be able to find some time at the beginning of next week during pyparis conference.

**Comment by Duologic** [comment]
No pressure, I've been busy too. 👍

File: django_statsd/clients/null.py [hunk 1]
```diff
@@ -1,8 +1,12 @@
+from __future__ import absolute_import
 from statsd.client import StatsClient
 
 
 class StatsClient(StatsClient):
     """A null client that does nothing."""
 
+    def __init__(self, host, port, prefix=None, **kwargs):
+        self._prefix = prefix
```
Comment (xordoquy) [inline django_statsd/clients/null.py:9]: What's the purpose of `_prefix` ?

File: django_statsd/middleware.py [hunk 2]
```diff
@@ -7,34 +8,38 @@
...
+except ImportError:  # pragma: no cover
+    MiddlewareMixin = object
+
+
+class GraphiteMiddleware(MiddlewareMixin):
 
     def process_response(self, request, response):
         statsd.incr('response.%s' % response.status_code)
-        if hasattr(request, 'user') and request.user.is_authenticated():
+        if hasattr(request, 'user') and request.user and request.user.is_authenticated():
```
Comment (xordoquy) [inline django_statsd/middleware.py:21]: Do we really have cases where request.user can be `None` ?

File: django_statsd/models.py [hunk 3]
```diff
@@ -1,11 +1,13 @@
...
+from django.contrib.auth.signals import user_logged_in, user_logged_out, user_login_failed
 from django_statsd.clients import statsd
 
-from .celery import register_celery_events
+from .celery_hooks import register_celery_events
 
 
-if getattr(settings, 'STATSD_CELERY_SIGNALS', False):
+if getattr(settings, 'STATSD_CELERY_SIGNALS', False):  # pragma: no cover
     register_celery_events()
```
Comment (xordoquy) [inline django_statsd/models.py:11]: Note to myself, this should go into app.py instead.

File: setup.py [hunk 4]
```diff
@@ -1,34 +1,55 @@
...
+    if req.link:
+        dependency_links.append(str(req.link))
+
+# Inject test requirements from requirements_test.txt into setup.py
+requirements_test_file = parse_requirements(path.join('requirements', 'requirements_test.txt'), session=PipSession())
+for req in requirements_test_file:
+    tests_require.append(str(req.req))
+    if req.link:
+        dependency_links.append(str(req.link))
+
```
Comment (xordoquy) [inline setup.py:26]: I'm not fan of this method. I'd really rather have a setup.py that doesn't depend on external files.

File: test_settings.py [hunk 5]
```diff
@@ -0,0 +1,43 @@
+DATABASES = {
```
Comment (xordoquy) [inline test_settings.py:1]: Not needed with the conftest.py

File: django_statsd/patches/__init__.py [hunk 6]
```diff
@@ -1,11 +1,12 @@
+from __future__ import absolute_import
 from django.conf import settings
+from importlib import import_module
 
-try:
-    from importlib import import_module
-except ImportError:
-    from django.utils.importlib import import_module
```
Comment (xordoquy) [inline django_statsd/patches/__init__.py:6]: :+1:

**Review by xordoquy — COMMENTED** [review]
Here's a raw review.

What's the reason to rename celery.py into celery_hooks.py ?
I would really keep tox because it has more flexibility than the run_test.sh script.
Somehow, it seems the travis file in this review looks more complex than the one I have.

File: django_statsd/clients/null.py (hunk 1 above)
Comment (Duologic) [reply django_statsd/clients/null.py:9]: StatsClient expects that the `_prefix` exists. The purpose of adding the __init__ is so it doesn't try to setup a socket connection using this client.

File: django_statsd/middleware.py (hunk 2 above)
Comment (Duologic) [reply django_statsd/middleware.py:21]: I think I cherry-pick't that one in. Figured it wouldn't harm checking if the user is actually something before calling an attribute.
The alternative would be that it throws an ugly 'NoneType doesn't have attribute is_authenticated' as we would assume that `request.user` can't be None.

**Comment by Duologic** [comment]
Replaced by https://github.com/django-statsd/django-statsd/pull/105

**Comment by Duologic** [comment]
_What's the reason to rename celery.py into celery_hooks.py ?_

This was due to this commit: https://github.com/django-statsd/django-statsd/commit/2ef4b1070b943378494271504ae7c4ad96a3ac64

> Stop shadowing the celery package
> This broke the celery.signals imports inside django_statsd.celery.

_I would really keep tox because it has more flexibility than the run_test.sh script._
_Somehow, it seems the travis file in this review looks more complex than the one I have._

Agreed, I've split the CI/test stuff off in a new pull request. If we want to change that, I think this should be done independently from the code changes.

**Pull request closed without merging**